core/primitives/src/trie_key.rs

```diff
@@ -0,0 +1,471 @@
+    DelayedReceiptIndices,
+    /// Used to store a delayed receipt `primitives::receipt::Receipt` for a given index `u64`
+    /// in a delayed receipt queue. The queue is unique per shard.
+    DelayedReceipt { index: u64 },
+    /// Used to store a key-value record `Vec<u8>` within a contract deployed on a given `AccountId`
+    /// and a given key.
+    ContractData { account_id: AccountId, key: Vec<u8> },
+}
+
+impl TrieKey {
+    pub fn len(&self) -> usize {
+        match self {
+            TrieKey::Account { account_id } => col::ACCOUNT.len() + account_id.len(),
+            TrieKey::ContractCode { account_id } => col::CONTRACT_CODE.len() + account_id.len(),
+            TrieKey::AccessKey { account_id, public_key } => {
+                col::ACCESS_KEY.len() * 2 + account_id.len() + public_key.len()
+            }
+            TrieKey::ReceivedData { receiver_id, data_id } => {
+                col::RECEIVED_DATA.len()
+                    + receiver_id.len()
+                    + ACCOUNT_DATA_SEPARATOR.len()
+                    + data_id.as_ref().len()
+            }
+            TrieKey::PostponedReceiptId { receiver_id, data_id } => {
+                col::POSTPONED_RECEIPT_ID.len()
+                    + receiver_id.len()
+                    + ACCOUNT_DATA_SEPARATOR.len()
+                    + data_id.as_ref().len()
+            }
+            TrieKey::PendingDataCount { receiver_id, receipt_id } => {
+                col::PENDING_DATA_COUNT.len()
+                    + receiver_id.len()
+                    + ACCOUNT_DATA_SEPARATOR.len()
+                    + receipt_id.as_ref().len()
+            }
+            TrieKey::PostponedReceipt { receiver_id, receipt_id } => {
+                col::POSTPONED_RECEIPT.len()
+                    + receiver_id.len()
+                    + ACCOUNT_DATA_SEPARATOR.len()
+                    + receipt_id.as_ref().len()
+            }
+            TrieKey::DelayedReceiptIndices => col::DELAYED_RECEIPT_INDICES.len(),
+            TrieKey::DelayedReceipt { .. } => col::DELAYED_RECEIPT.len() + size_of::<u64>(),
+            TrieKey::ContractData { account_id, key } => {
+                col::ACCOUNT.len() + account_id.len() + ACCOUNT_DATA_SEPARATOR.len() + key.len()
+            }
+        }
+    }
+
+    pub fn to_vec(&self) -> Vec<u8> {
+        let expected_len = self.len();
+        let mut res = Vec::with_capacity(expected_len);
+        match self {
+            TrieKey::Account { account_id } => {
+                res.extend(col::ACCOUNT);
+                res.extend(account_id.as_bytes());
+            }
+            TrieKey::ContractCode { account_id } => {
+                res.extend(col::CONTRACT_CODE);
+                res.extend(account_id.as_bytes());
+            }
+            TrieKey::AccessKey { account_id, public_key } => {
+                res.extend(col::ACCESS_KEY);
+                res.extend(account_id.as_bytes());
+                res.extend(col::ACCESS_KEY);
```

Comment:
Should it be `ACCOUNT_DATA_SEPARATOR`?

Reply:
It's legacy logic.

Reply:
So the reason why it does not crash is because `ACCESS_KEY == [2]` and `'\0x2'` is not a valid character for `account_id`.

Reply:
Yep